dbms/src/Core/DecimalFunctions.h

```diff
@@ -0,0 +1,135 @@
+#pragma once
+// Moved Decimal-related functions out from Core/Types.h to reduce compilation time.
+
+#include <Core/Types.h>
+#include <Common/intExp.h>
+
+#include <limits>
+
+class DateLUTImpl;
+
+namespace DB
+{
+
+static constexpr size_t minDecimalPrecision() { return 1; }
+template <typename T> static constexpr size_t maxDecimalPrecision() { return 0; }
+template <> constexpr size_t maxDecimalPrecision<Decimal32>() { return 9; }
+template <> constexpr size_t maxDecimalPrecision<Decimal64>() { return 18; }
+template <> constexpr size_t maxDecimalPrecision<Decimal128>() { return 38; }
+
+template <typename T> T decimalScaleMultiplier(UInt32 scale);
+template <> inline Int32 decimalScaleMultiplier<Int32>(UInt32 scale) { return common::exp10_i32(scale); }
+template <> inline Int64 decimalScaleMultiplier<Int64>(UInt32 scale) { return common::exp10_i64(scale); }
+template <> inline Int128 decimalScaleMultiplier<Int128>(UInt32 scale) { return common::exp10_i128(scale); }
+
+template <typename T>
+struct DecimalComponents
+{
+    T whole;
+    T fractional;
+};
+
+template <typename DecimalType>
+DecimalType decimalFromComponentsWithMultiplier(const typename DecimalType::NativeType & whole,
+                                                 const typename DecimalType::NativeType & fractional,
+                                                 typename DecimalType::NativeType scale_multiplier)
+{
+    using T = typename DecimalType::NativeType;
+    const T value = whole * scale_multiplier + fractional % scale_multiplier;
+    return DecimalType(value);
+}
+
+template <typename DecimalType>
+DecimalType decimalFromComponents(const typename DecimalType::NativeType & whole, const typename DecimalType::NativeType & fractional, UInt32 scale)
+{
+    using T = typename DecimalType::NativeType;
+
+    return decimalFromComponentsWithMultiplier<DecimalType>(whole, fractional, decimalScaleMultiplier<T>(scale));
+}
+
+template <typename DecimalType>
+DecimalType decimalFromComponents(const DecimalComponents<typename DecimalType::NativeType> & components, UInt32 scale)
+{
+    return decimalFromComponents<DecimalType>(components.whole, components.fractional, scale);
+}
+
+template <typename DecimalType>
+DecimalComponents<typename DecimalType::NativeType> decimalSplitWithScaleMultiplier(const DecimalType & decimal, typename DecimalType::NativeType scale_multiplier)
+{
+    using T = typename DecimalType::NativeType;
+    const auto whole = decimal.value / scale_multiplier;
+    auto fractional = decimal.value % scale_multiplier;
+    if (fractional < T(0))
```

Comment:
And this code says that for `value=-11`, `scale=10` the result will be `{-1, 1}`